### ISSUE TITLE
Connection Nodes should resolve earlier

### ISSUE THREAD

**Issue by jasonbahl** [wp-graphql/wp-graphql#1138]
Nodes are resolved too late by connection resolvers. 

What I mean by this, is that currently connection resolvers query for IDs of nodes, and pass them down through edges, to nodes, and at the node level is when the ID is passed to a Deferred resolver to resolve the node. The deferred resolver allows for many nodes to be resolved at once because of the DataLoader pattern. 

The problem with this, is that the nodes are being resolved too late to make decisions on their privacy. 

Take a query like the following:

```
{
  users {
    nodes {
       id
       name
    }
  }
}
```

A public user should get a list of users. Only users that have published posts are considered public, but that's determined after the User has been queried for. We take that user, check if it has published posts, if it does, it's public, if not, it's not public and a public user shouldn't have said user in the connection. 

Because we resolve the nodes so late, we end up with errors, something like the following: 

`Cannot return null for non-nullable field User.id.`

Also, when registering edge fields, the `$root` that is passed down to the resolver doesn't have the node as context, just the ID of the node. This can make it difficult to resolve edge fields properly when actual node data is often necessary, not just an ID. 

## Proposal
The nodes of a connection should be resolved and passed through the Model Layer _before_ the list of edges/nodes is  passed down to the next level of the tree. This will allow for the nodes that are private or otherwise considered not-visible for the requesting user to be stripped from the results prior to passing down to edges/nodes. 

This will solve the `id cannot be null` errors while also providing more rich context to edge fields, and performance will not be negatively impacted as we can still make use of deferred resolvers, just at a higher point in the resolve tree.

**Comment by jasonbahl** [comment]
Here's a more concrete example illustrating the above: 

Let's say we're dealing with something like a membership plugin that uses meta to flag what level of member has access to a Post. 

The Model Layer could be filtered like so, to determine whether the Post should be considered private: 

```
add_filter( 'graphql_data_is_private', function( $is_private, $model_name, $data, $visibility, $owner, $current_user ) {
	if ( 'PostObject' === $model_name ) {
		$is_private_meta = get_post_meta( $data->ID, 'is_private' );
		if ( isset( $is_private_meta ) && true === (bool) $is_private_meta ) {
			$is_private = true;
		}
	}
	return $is_private;
}, 10, 6 );
```

Let's add the `is_private` meta to a Post, then make a query for posts. 

`update_post_meta( $post_id, 'is_private', true );`

Now, let's query for a connection of posts: 

```
{
  posts {
    nodes {
      id
      postId
      title
    }
  }
}
```

![Screen Shot 2020-01-30 at 7 02 37 AM](https://user-images.githubusercontent.com/1260765/73456113-8a88d000-432e-11ea-94c1-a50e9af3bcc8.png)

We get the following error and a null node in the results. 😱 

This happens because the Nodes are resolved at the level of the tree in which they are returned. 

The above query resolves like so: 

```
{
  posts { # Call WP_Query for ONLY IDs. Pass the IDs to edges/nodes.
     nodes { # Call the resolve_node method, which is a deferred resolver and resolves immediately before returning
        id . # Return the node ID, but since the node is private (null) there's no ID, and we have this error
        postId
        title
     }
  }
}
```
What execution should change to is: 

```
{
  posts { # Call WP_Query for ONLY IDs. Pass the IDs to a deferred "resolveMany" function to resolve all nodes at this level of the tree from DataLoader. Filter out any null nodes from the resolved. Pass the filtered list of nodes to edges/nodes.
     nodes { # Resolve using the node object passed in
        id . # Return the node ID
        title
     }
  }
}
```

**Comment by jasonbahl** [comment]
closed by #1193
